packages/testing/src/chai-retry-plugin/chai-retry-plugin.ts

```diff
@@ -0,0 +1,187 @@
+ * ```
+ * @example
+ * ```ts
+ * await expect(sometimesNullFunction).retry({ timeout: 2000 }).to.be.not.null;
+ * ```
+ * @example
+ * ```ts
+ * await expect(funcToRetry).retry({ retries: 5, delay: 10 }).and.have.property('success').and.be.true;
+ * ```
+ */
+export const chaiRetryPlugin = function (_: typeof Chai, utils: Chai.ChaiUtils) {
+    Assertion.addMethod('retry', function (retryOptions: RetryOptions = {}): PromiseLikeAssertion {
+        const functionToRetry: FunctionToRetry = this._obj as FunctionToRetry;
+
+        if (typeof functionToRetry !== 'function') {
+            throw new TypeError(utils.inspect(functionToRetry) + ' is not a function.');
+        }
+
+        const assertionStack: AssertionStackItem[] = [];
+        const defaultRetryOptions: Required<RetryOptions> = { timeout: 5000, retries: Infinity, delay: 0 };
+        const options: Required<RetryOptions> = { ...defaultRetryOptions, ...retryOptions };
+
+        // Fake assertion object for catching calls of chained methods
+        const proxyTarget = new Assertion({});
+
+        /**
+         * This is needed to handle cases when assertion ends with property, for example:
+         * await expect(func).retry().to.be.null;
+         */
+        const assertionPropertyKeys = [
+            'ok',
+            'true',
+            'null',
+            'false',
+            'undefined',
+            'empty',
+            'NaN',
+            'finite',
+            'exist',
+            'arguments',
+            'all',
+            'own',
+        ];
+
+        const assertionProxy: PromiseLikeAssertion = Object.assign(
+            new Proxy(proxyTarget, {
+                get: function (target: Chai.Assertion, key: string) {
+                    if (assertionPropertyKeys.includes(key)) {
+                        assertionStack.push({ key });
+
+                        return assertionProxy;
+                    }
+
+                    if (key === 'not') {
+                        assertionStack.push({ isNegate: true });
+
+                        return assertionProxy;
+                    }
+
+                    // Handle native Chai's assertion methods and 'then' call
+                    const value = target[key as keyof Chai.Assertion];
+
+                    if (typeof value === 'function') {
+                        return (...args: unknown[]) => {
+                            if (key === 'then') {
+                                return (value as unknown as AssertionMethod)(...args);
+                            }
+
+                            assertionStack.push({ method: value as unknown as AssertionMethod, args });
+
+                            return assertionProxy;
+                        };
+                    }
+
+                    return assertionProxy;
+                },
+            }),
+            {
+                then: (resolve: () => void, reject: () => void) => {
+                    return getRetryPromiseWithTimeout({
```

Comment:
The main task of this function is not returning `retryPromise`, but actually retrying function (with assertions) and it should be name respectively. The fact that it returns a promise is just an implementation aspect.

Reply:
@yurii-ve fixed